src/common/slow-query/src/recorder.rs

```diff
@@ -0,0 +1,355 @@
+use session::context::{QueryContextBuilder, QueryContextRef};
+use snafu::ResultExt;
+
+use crate::error::{CatalogSnafu, CreateTableSnafu, InsertRowsSnafu, Result, TableNotFoundSnafu};
+
+const SLOW_QUERY_TABLE_NAME: &str = "slow_queries";
+const SLOW_QUERY_TABLE_COST_COLUMN_NAME: &str = "cost";
+const SLOW_QUERY_TABLE_THRESHOLD_COLUMN_NAME: &str = "threshold";
+const SLOW_QUERY_TABLE_QUERY_COLUMN_NAME: &str = "query";
+const SLOW_QUERY_TABLE_IS_PROMQL_COLUMN_NAME: &str = "is_promql";
+const SLOW_QUERY_TABLE_TIMESTAMP_COLUMN_NAME: &str = "timestamp";
+
+/// SlowQueryRecorder is responsible for recording slow queries.
+#[derive(Clone)]
+pub struct SlowQueryRecorder {
+    slow_query_opts: SlowQueryOptions,
+    inserter: InserterRef,
+    statement_executor: StatementExecutorRef,
+    catalog_manager: CatalogManagerRef,
+}
+
+impl SlowQueryRecorder {
+    /// Create a new SlowQueryRecorder.
+    pub fn new(
+        slow_query_opts: SlowQueryOptions,
+        inserter: InserterRef,
+        statement_executor: StatementExecutorRef,
+        catalog_manager: CatalogManagerRef,
+    ) -> Self {
+        Self {
+            slow_query_opts,
+            inserter,
+            statement_executor,
+            catalog_manager,
+        }
+    }
+
+    /// Start a new SlowQueryTimer. Return `None` if the `slow_query.enable` is false.
+    /// When the query is finished, you need to call `stop()` to record the slow query.
+    pub fn start(
+        &self,
+        stmt: QueryStatement,
+        query_ctx: QueryContextRef,
+    ) -> Option<SlowQueryTimer> {
+        if self.slow_query_opts.enable {
+            Some(SlowQueryTimer {
+                start: std::time::Instant::now(),
+                stmt,
+                threshold: self.slow_query_opts.threshold,
+                sample_ratio: self.slow_query_opts.sample_ratio,
+                query_ctx,
+                inserter: self.inserter.clone(),
+                statement_executor: self.statement_executor.clone(),
+                catalog_manager: self.catalog_manager.clone(),
+                record_type: self.slow_query_opts.record_type,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+/// SlowQueryTimer is used to log slow query when it's dropped.
+pub struct SlowQueryTimer {
+    start: std::time::Instant,
+    stmt: QueryStatement,
+    threshold: Option<Duration>,
+    sample_ratio: Option<f64>,
+    query_ctx: QueryContextRef,
+    inserter: InserterRef,
+    statement_executor: StatementExecutorRef,
+    catalog_manager: CatalogManagerRef,
+    record_type: SlowQueriesRecordType,
+}
+
+impl SlowQueryTimer {
+    /// Stop the slow query timer and record the slow query.
+    pub async fn stop(&self) {
+        if self.record_type == SlowQueriesRecordType::SystemTable {
+            if let Err(e) = self.create_system_table().await {
+                error!(e; "Failed to create system table for slow queries");
+                return;
+            }
+        }
```

Comment:
Better move table creation to `insert_slow_query` to ensure that slow-query table is really lazy.